format.md

```diff
@@ -0,0 +1,138 @@
+# About this document
+
+This document attempts to document the format currently implemented by binjs-fbssdc, also known as "context 0.1".
+
+# Global structure
+
+```
+Stream ::= MagicHeader BrotliBody
+```
+
+# Magic Header
+
+The magic header serves to identify the format.
+
+```
+MagicHeader ::= "\x89BJS\r\n\0\n" FormatVersion
+FormatVersion ::= 0b00000010
+```
+
+# Brotli content
+
+With the exception of the header, the entire file is brotli-compressed.
+
+```
+BrotliBody ::= Brotli(Body)
+Body ::= Prelude AST
+```
+
+Where `Brotli(...)` represents data that may be uncompressed by the
+`brotli` command-line tool or any compatible library.
+
+# Prelude
+
+The prelude defines a dictionary of strings and a dictionary of probabilities. The order of both is meaningful.
+
+```
+Prelude ::= StringPrelude ProbabilityPrelude
+```
+
+## String dictionary
+
+```
+StringPrelude ::= n=NumberOfStrings StringDefinition{n}
+NumberOfStrings ::= varnum
+StringDefinition ::= NonZeroByte* ZeroByte
+NonZeroByte ::= 0x01-0xFF
+ZeroByte ::= 0x00
+```
+
+Strings are utf-8 encoded, then we replace any embedded `0x00 0x01` with `0x01 0x01` (FIXME: Why does this work?)
+
+
+
+## Probability dictionaries
+
+```
+ProbabilityPrelude ::= ProbabilityTable* # FIXME: How do we determine the number of tables?
+ProbabilityTable ::= ProbabilityTableUnreachable     # Compression artifact. A table that needs to appear here but is never used.
+                 |   ProbabilityTableOptimizedOne    # Optimization: A probability table with a single symbol.
+                 |   ProbabilityTableExplicitSymbols # Used for strings, numbers.
+                 |   ProbabilityTableIndexedSymbols  # Used for enums, booleans, sums of interfaces.
+```
+
+The order of probability tables is meaningful.
+
+```
+ProbabilityTableUnreachable ::= 0x02
+ProbabilityTableOptimizedOne ::= 0x00 ExplicitSymbolData # Used for strings, numbers.
+                             |   0x00 Index              # Used for enums, booleans, sums of interfaces.
+ProbabilityTableExplicitSymbols ::= 0x01 n=ProbabilityTableLen Probability{n} ExplicitSymbolData{n} # Only list the symbols actually used.
+ProbabilityTableIndexedSymbols  ::= 0x01 Probability* # List all symbols, in the order extracted from the grammar.
+ProbabilityTableLen ::= varnum
+Index ::= varnum
+Probability ::= u8
```

Comment:
to my understanding, this is not probability but the length of the code used in tree.
when there are 2 symbols possible, one gets `code=0` and one gets `code=1`, so the length of both code are `1`,
so only 2 items get `length=1` and others get `length=0`.

when there are 3 symbols possible, one gets `code=0`, one gets `code=10`, and one gets `code=11`.
so the lengths are `1`, `2`, `2` for each.

of course the length implies the possibility (shorter is higher) tho.

Reply:
Ah, my bad, you're right.